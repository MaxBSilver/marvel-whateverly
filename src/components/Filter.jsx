import React, { Component } from 'react'

export class Filter extends Component {
  constructor() {
    super();
    this.state = {
      movies: false,
      comics: false
    }
  }

  handleComics = e => {
    e.preventDefault();
    e.target.id === 'comics'
      ? this.setState({comics: !this.state.comics})
      : this.setState({movies: false})
    if (this.state.movies) this.setState({movies: false});
    this.props.rendered.push(...[this.props.comics])
  }

  handleMovies = e => {
    e.preventDefault();
    e.target.id === 'movies'
      ? this.setState({movies: !this.state.movies})
      : this.setState({comics: false})
    if (this.state.comics) this.setState({comics: false});
    this.props.rendered.push(...[this.props.movies])
  }

  // object.values over current movies/comics, index 0
  // object.keys over that and map into a dropdown

  // filterData = () => {
  //   let val = Object.values(this.props.movies)
  //   val[0].map(v => {
  //     return <option>{v}</option>
  //   })
  // }

  // switch(true) {
  //   case this.state.movies:
  //   Object.values(this.props.movies)
  //     val[0].map(v => {
  //       return <option>{v}</option>
  //     })
  //     break;
  //   case this.state.comics:
  //     break;
  //   default:
  // }

  
  render() {

    let val = Object.values(this.props.movies)
    let dropDownMovies = Object.keys(val).map(v => {
    return (<option>{v}</option>)})
    let com = Object.values(this.props.comics)
    let dropDownComics = Object.keys(com).map(v => {
    return (<option>{v}</option>)})
              
    switch(true) {
      case this.state.movies:
      return (
            <div>
            <select>
              <option selected="selected">--Filter by--</option>
            {dropDownMovies}
            </select>
            </div>
            )
      break;
      case this.state.comics:
      return (
        <div>
        <select>
          <option selected="selected">--Filter by--</option>
        {dropDownComics}
        </select>
        </div>
        )
      break;
      default:
      // return false
    
      return (
      <form >
        <input onClick={this.handleComics} type="submit" value="Comics" id="comics"/>
        <input onClick={this.handleMovies} type="submit" value="Movies" id="movies"/>
      </form>
      )
    }
  }
}

export default Filter
