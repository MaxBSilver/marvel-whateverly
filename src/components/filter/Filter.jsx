import React, { Component } from 'react'

export class Filter extends Component {
  constructor() {
    super();
    this.state = {
      movies: false,
      comics: false,
      favorites: false
    }
  }

  handleComics = e => {
    e.preventDefault();
    e.target.id === 'comics'
    ? this.setState({comics: !this.state.comics})
    : this.setState({movies: false})
    if (this.state.movies) this.setState({movies: false});
    this.props.storeRendered(...[Object.values(this.props.comics)])
    this.props.setSearchDataset(Object.values(this.props.comics))
  }

  handleMovies = e => {
    e.preventDefault();
    e.target.id === 'movies'
    ? this.setState({movies: !this.state.movies})
    : this.setState({comics: false})
    if (this.state.comics) this.setState({comics: false});
    this.props.storeRendered(...[Object.values(this.props.movies)])
    this.props.setSearchDataset(Object.values(this.props.movies))
  }

  handleFavorites = e => {
    e.preventDefault();
    this.state.favorites
    ? this.props.storeRendered(this.props.data.slice(0, 10))
    : this.props.storeRendered(JSON.parse(localStorage.getItem('marvelous')));
    this.setState({favorites: !this.state.favorites});
  }

  resetSearch = () => {
    this.props.setSearchDataset(this.props.data);
    this.props.storeRendered(this.props.data);
    this.setState({comics: false, movies: false})
  }
  
  render() {
    let mov = Object.values([this.props.movies][0])[0]
    let com = Object.values([this.props.comics][0])[0]
    let excludedFilters = ['title', 'link', 'img', 'id', 'favorite']
    switch(true) {
      case this.state.movies:
      return (
        <div className="filter-form">
          <button className="filter-btn" onClick={this.resetSearch} type="button">
            Show All
          </button>
          <input className="filter-btn" onClick={this.handleComics} type="submit" value="Show Comics" id="comics"/>
          <select className="filter-dropdown">
            <option value="" selected="selected">--Filter by--</option>
          {Object.keys(mov).map(m => excludedFilters.includes(m) ? true : (<option value={m}>{m}</option>))}
          </select>
        </div>
      )
      case this.state.comics:
      return (
        <div className="filter-form">
          <button className="filter-btn" onClick={this.resetSearch} type="button">
              Show All
          </button>
          <input className="filter-btn" onClick={this.handleMovies} type="submit" value="Show Movies" id="movies"/>
          <select className="filter-dropdown">
            <option value="" selected="selected">--Filter by--</option>
            {Object.keys(com).map(c => excludedFilters.includes(c) ? true : (<option value={c}>{c}</option>))}
          </select>
        </div>
      )
      default:
      return (
      <form >
        <input className="filter-btn" onClick={this.handleFavorites} type="submit" 
        value={this.state.favorites ? "Show Random": "Show Favorites"} id="comics"/>
        <input className="filter-btn" onClick={this.handleComics} type="submit" value="Comics" id="comics"/>
        <input className="filter-btn" onClick={this.handleMovies} type="submit" value="Movies" id="movies"/>
      </form>
      )
    }
  }
}

export default Filter
