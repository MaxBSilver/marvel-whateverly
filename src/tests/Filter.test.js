import React from 'react';
import Filter from '../components/filter/Filter';
import { shallow } from 'enzyme';

describe('Filter', () => {
  
  const mockStoreRendered = jest.fn();
  
  const mockSetSearchDataset = jest.fn();
  
  let wrapper;
  let mockSortData = [
    {mockNum: 5}, 
    {mockNum: 4}, 
    {mockNum: 3}, 
    {mockNum: 6}, 
    {mockNum: 7}, 
    {mockNum: 1}, 
    {mockNum: 2}
  ]
    
  let mockMovies = {
    "title": "Deadpool",
    "releaseYear": 2016,
    "directors": ["Tim Miller"],
    "rating": "R",
    "stars": ["Ryan Reynolds", "Morena Baccarin", "T.J. Miller", "Ed Skrein", "Brianna Hildebrand"],
    "imdbRating": 8,
    "characters": ["Deadpool", "Negasonic Teenage Warhead", "Ajax", "Colossus"],
    "basedOn": ["Deadpool Vol. 1 #1"],
    "link": "https://www.imdb.com/title/tt1431045/",
    "img": "https://images-na.ssl-images-amazon.com/images/I/51Gh9OaWVcL.jpg",
    "id": 11,
    "favorite": false
  }

  let mockComics = {
    "title": "Civil War",
    "publishDate": 2006,
    "writers": ["Mark Millar"],
    "pencillers": ["Steve McNiven"],
    "inkers": ["Dexter Vines"],
    "letterers": ["Chris Eliopoulos"],
    "colorists": ["Morry Hollowell"],
    "editors": ["Molly Lazer", "Aubrey Sitterson", "Andy Schmidt", "Tom Brevoort"],
    "characters": ["Iron Man", "Spider-Man", "Doctor Strange", "Luke Cage", "Mister Fantastic", "Henry Pym", "Captain America", "Black Panther"],
    "adapted": true,
    "summary": "After a huge number of civilian lives lost, the government tries to enact The Superhero Registration Act, an authoritarian law where all heroes will have to register with the government. Some heroes, like Iron Man, are in favor of the law while others, like Captain America, are not.",
    "link": "https://marvel.fandom.com/wiki/Civil_War_(Event)",
    "img": "https://vignette.wikia.nocookie.net/marveldatabase/images/0/0f/Civil_War_Vol_1_1_Turner_Variant.jpg/revision/latest?cb=20070524065830",
    "id": 15,
    "favorite": false
  }

  beforeEach(() => {
    wrapper = shallow(
      <Filter movies={mockSortData} comics={mockComics} storeRendered={mockStoreRendered} setSearchDataset={mockSetSearchDataset} />
    )
  });

  it('should have a proper default state', () => {
    expect(wrapper.state()).toEqual( {movies: false, comics: false, favorites: false} );
  });

  it('should match the snapshot with all the data passed in', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should filter comics', () => {
    wrapper.setState( {comics: mockComics} )
    wrapper.instance().handleComics( { target: {id: 'comics'}, preventDefault: () => {} } )
    expect(wrapper.state('comics')).toEqual( false );
    wrapper.instance().handleComics( { target: {id: 'comics'}, preventDefault: () => {} } )
    expect(wrapper.state('comics')).toEqual( true );
    expect(mockStoreRendered).toHaveBeenCalled();
    expect(mockSetSearchDataset).toHaveBeenCalled();
  });

  it('should filter movies', () => {
    expect(wrapper.state()).toEqual( { comics: false, movies: false, favorites: false } )
    wrapper.setState( {comics: mockMovies} )
    wrapper.instance().handleMovies( { target: {id: 'movies' }, preventDefault: () => {} } )
    expect(wrapper.state()).toEqual( { comics: false, movies: true, favorites: false } );
  });

  it('should filter favorites', () => {
    expect(wrapper.state()).toEqual( { comics: false, movies: false, favorites: false } )
    wrapper.instance().handleFavorites( { target: {id: 'favorites' }, preventDefault: () => {} } )
    expect(wrapper.state()).toEqual( { comics: false, movies: false, favorites: true } );
  });

  it('should reset the search', () => {
    wrapper.instance().resetSearch()
    expect(mockSetSearchDataset).toHaveBeenCalled()
    expect(mockStoreRendered).toHaveBeenCalled()
    expect(wrapper.state()).toEqual( { comics: false, movies: false, favorites: false } );
  });

  it('should sort cards based on option', () => {
    wrapper.instance().setState({movies: true})
    wrapper.instance().sortCards({target: {value: 'mockNum'}})
    expect(mockStoreRendered).toHaveBeenCalledWith([
      {mockNum: 7}, 
      {mockNum: 6}, 
      {mockNum: 5}, 
      {mockNum: 4}, 
      {mockNum: 3}, 
      {mockNum: 2}, 
      {mockNum: 1}
    ])
  })
})