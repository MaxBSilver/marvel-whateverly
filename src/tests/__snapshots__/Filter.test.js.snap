// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter should match the snapshot with all the data passed in 1`] = `
<form>
  <input
    className="filter-btn"
    id="comics"
    onClick={[Function]}
    type="submit"
    value="Show Favorites"
  />
  <input
    className="filter-btn"
    id="comics"
    onClick={[Function]}
    type="submit"
    value="Comics"
  />
  <input
    className="filter-btn"
    id="movies"
    onClick={[Function]}
    type="submit"
    value="Movies"
  />
</form>
`;
